write-fonts/src/tables/cmap.rs

```diff
@@ -105,27 +74,148 @@ impl Cmap {
+        let mut start_glyph_id = cmap[&start_char_code];
+        let mut last_glyph_id = start_glyph_id.wrapping_sub(1);
+        let mut last_char_code = start_char_code.wrapping_sub(1);
+        let mut groups = Vec::new();
+        for char_code in char_codes {
+            let glyph_id = cmap[&char_code];
+            if glyph_id != last_glyph_id.wrapping_add(1)
+                || char_code != last_char_code.wrapping_add(1)
+            {
+                groups.push((start_char_code, last_char_code, start_glyph_id));
+                start_char_code = char_code;
+                start_glyph_id = glyph_id;
+            }
+            last_glyph_id = glyph_id;
+            last_char_code = char_code;
+        }
+        groups.push((start_char_code, last_char_code, start_glyph_id));
+
+        let num_groups: u32 = groups.len().try_into().unwrap();
+        let seq_map_groups = groups
+            .into_iter()
+            .map(|(start_char, end_char, gid)| SequentialMapGroup::new(start_char, end_char, gid))
+            .collect::<Vec<_>>();
+        CmapSubtable::format_12(
+            size_of_cmap12(num_groups),
+            0, // 'lang' set to zero for all 'cmap' subtables whose platform IDs are other than Macintosh
+            num_groups,
+            seq_map_groups,
+        )
+    }
+}
+
+impl Cmap {
+    /// Generates a [cmap](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap) that is expected to work in most modern environments.
+    ///
+    /// This emits [format 4](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-4-segment-mapping-to-delta-values)
+    /// and [format 12](https://learn.microsoft.com/en-us/typography/opentype/spec/cmap#format-12-segmented-coverage)
+    /// subtables, respectively for the Basic Multilingual Plane and Full Unicode Repertoire.
+    pub fn from_mappings(mappings: impl IntoIterator<Item = (char, GlyphId)>) -> Cmap {
+        let mut mappings: Vec<_> = mappings.into_iter().collect();
+        mappings.sort();
+
+        let mut uni_records = Vec::new(); // platform 0
+        let mut win_records = Vec::new(); // platform 3
+
+        let bmp_subtable = CmapSubtable::create_format_4(&mappings);
+        if let Some(bmp_subtable) = bmp_subtable {
+            // Absent a strong signal to do otherwise, match fontmake/fonttools
+            // Since both Windows and Unicode platform tables use the same subtable they are
+            // almost entirely byte-shared
+            // See https://github.com/googlefonts/fontmake-rs/issues/251
+            uni_records.push(EncodingRecord::new(
+                PlatformId::Unicode,
+                UNICODE_BMP_ENCODING,
+                bmp_subtable.clone(),
+            ));
+            win_records.push(EncodingRecord::new(
+                PlatformId::Windows,
+                WINDOWS_BMP_ENCODING,
+                bmp_subtable,
+            ));
+        }
+
+        if mappings.iter().any(|(cp, _)| *cp > '\u{FFFF}') {
```

Comment:
`> 0xFFFF` is meaningless unless you know unicode; I suggest a function is-bmp, perhaps as a trait similar to ot_round so this becomes something like `.any(|(cp, _)| !cp.is_basic_multilingual_plan())`

Reply:
I'll add a comment. I don't think that trait would be used much except here for cmap